### ISSUE TITLE
tensorflow 2.15 "Illegal instruction: 4" Mac M1 processor

### ISSUE THREAD

**Issue by skwyddie** [tensorflow/tensorflow#62860]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.15.0

### Custom code

No

### OS platform and distribution

Darwin MacBook-Pro-2.localdomain 23.3.0 Darwin Kernel Version 23.3.0; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000 arm64 (Sonoma 14.3 Apple M1 Max)

### Python version

3.11.7

### Current behavior?

$ python3
Python 3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow
Illegal instruction: 4
$

I expected to be able to import the module after it is installed. I suspect the newer processor is an issue.

### Standalone code to reproduce the issue

```shell
$ python3
Python 3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow
Illegal instruction: 4
$
```

**Comment by sushreebarsa** [comment]
@skwyddie  ensorFlow 2.15 is not natively compatible with M1 so you need to switch to a version built for Apple Silicon. TensorFlow 2.10.0-rc0 is the latest release with official M1 support and TensorFlow 2.9.1 is a more stable option, also offering M1 compatibility. Please refer to the TensorFlow M1 installation guide for detailed instructions: https://developer.apple.com/metal/tensorflow-plugin/

Thank you!

**Comment by slashdottir** [comment]
Same problem happens with TensorFlow 2.10.0-rc0 on Mac M1 Pro

**Comment by skwyddie** [comment]
I can't find any version of tf that works on M1. I've tried everything down to 2.9.1.

**Comment by sushreebarsa** [comment]
@skwyddie You're right, there isn't an official TensorFlow version compiled specifically for Apple M1 yet. However, there are workarounds and alternative options you can explore. Here's the guide to install TensorFlow Metal: https://github.com/apple/tensorflow_macos
However, note that it's still under development and might have limitations or compatibility issues.
Thank you!

**Comment by thesofakillers** [comment]
This issue should be reopened, I am still facing the same issue, even with 2.16

**Comment by Konippi** [comment]
I'm also facing the same problem with tensorflow 2.16. (M2 pro, Sonoma14.5)
When I try to run `import tensorflow`, I get `illegal instruction 4`.
How can I solve it...

**Comment by thesofakillers** [comment]
fyi, check whether your python was installed via conda, and whether the conda installed is actually the ARM version of conda and not the x86 version of conda. I was mistakingly using x86 conda after transferring over from an old intel-based mac. Changing conda to the correct architecture fixed this issue for me.

**Comment by Konippi** [comment]
I have installed rye, so use venv as a virtual environment now.
Does anyone know how to solve this case?

**Comment by thesofakillers** [comment]
the issue i mention above is not conda exclusive fwiw. 

Try running

```console
python -c "import platform; print(platform.machine())"
```

This should print 

```
arm64
```

If you're on apple silicon. If it isn't and you are on apple silicon, something about your python installation is wrong, and tensorflow won't work.

**Comment by sahiljhawar** [comment]
On my local arm Mac machine it prints `x86_64`. The Conda environment I used to test this is setup using an official guide given on Anaconda's TF documentation

**Comment by sahiljhawar** [comment]
@thesofakillers Thanks, it worked. The trick is to use correct architecture:

`CONDA_SUBDIR=osx-arm64 conda create -n tf -c conda-forge python=3.11`

`CONDA_SUBDIR=osx-arm64` is important for arm Mac

**Comment by xunsheng** [comment]
Try 
conda config --set subdir osx-arm64 
If CONDA_SUBDIR=osx-arm64 doesn't work.
